### PR TITLE
Fixes a bug with the scope getting stuck on mobs that get dusted or ghost

### LINKED ISSUE #24189: getting dusted while scoped in leaves you in a bugged state
Reporting client version: 514.1589

## Round ID:
9598

## Testmerges:
- [Cargo Engineering Content Part 1 | My little machines that go wwwhrhhhr | Die of CO2 poisoning](https://github.com/Skyrat-SS13/Skyrat-tg/pull/24058)
- [URL Clickabillity. OOC Notes/Flavor Text/Species lore](https://github.com/Skyrat-SS13/Skyrat-tg/pull/24068)
- [[READY] [MODULAR] Blueshift wakes up after taking "Worst Nap Ever" [MDB IGNORE]](https://github.com/Skyrat-SS13/Skyrat-tg/pull/23604)

## Reproduction:
1. use the scope on an MCR or similar weapon
2. get dusted (in this example, by dying as marksman in CTF)
3. be put in this state

mouse is not visible, but can still interact with objects; character is offset from screen center, and returns as soon as you unscope (which requires being living)

image

### LINKED ISSUE #78756: Vision can get stuck scoped
Reporting client version: 514.1589

## Round ID:
[215998](https://scrubby.melonmesa.com/round/215998)

## Reproduction:
take a rifle that can be used to scope in (in my case the ctf marksman's rifle), then get dusted and as a ghost your camera will be stuck as if you were scoped in the direction you were scoped in, your cursor is also made invisible.

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

If you ghost while mid-scope (or similarly, if you get dusted/your mob gets deleted) the mob's cursor would get all messed up due to some oversights. This just clears up some of these logic errors and ensures everything gets cleaned up properly.

Fixes https://github.com/Skyrat-SS13/Skyrat-tg/issues/24189
Fixes https://github.com/tgstation/tgstation/issues/78756

## Why It's Good For The Game

Fixes a bug. I don't think scopes are even in use here but they are downstream.

<details>
<summary>Deleted and being dusted resets perspective and cursor properly</summary>

dreamseeker_SSU4YnEK4n

dreamseeker_GZEhuSJSGS

</details>

<details>
<summary>So does ghosting</summary>

dreamseeker_7n3uXZvDSI

</details>

## Changelog

:cl:
fix: being killed or ghosting while being scoped will no longer cause the cursor offset to persist in a bugged state
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
Fixes #78756?

File: code/datums/components/scope.dm [hunk 1]
```diff
@@ -101,21 +105,36 @@
...
+	if(zoom(user))
+		zooming = TRUE
+
+/**
+ * We start zooming by hiding the mouse pointer, adding our tracker overlay and starting our processing.
+ *
+ * Arguments:
+ * * user: The mob we are starting zooming on.
+*/
+/datum/component/scope/proc/zoom(mob/living/user)
```
Comment (san7890) [inline code/datums/components/scope.dm:123]: why is this on the `living` level? feels like a hurtful typecast when we aren't even sure that it is a living mob or not.

File: code/datums/components/scope.dm [hunk 2]
```diff
@@ -101,21 +105,36 @@
...
+	if(zooming)
+		return
 	user.client.mouse_override_icon = 'icons/effects/mouse_pointers/scope_hide.dmi'
 	user.update_mouse_pointer()
 	user.playsound_local(parent, 'sound/weapons/scope.ogg', 75, TRUE)
 	tracker = user.overlay_fullscreen("scope", /atom/movable/screen/fullscreen/cursor_catcher/scope, 0)
 	tracker.assign_to_mob(user, range_modifier)
+	tracker_owner_ckey = user.ckey
 	RegisterSignal(user, COMSIG_MOB_SWAP_HANDS, PROC_REF(stop_zooming))
+	RegisterSignal(user, COMSIG_QDELETING, PROC_REF(stop_zooming))
```
Comment (san7890) [inline code/datums/components/scope.dm:135]:
```suggestion
	RegisterSignals(user, list(COMSIG_MOB_SWAP_HANDS, COMSIG_QDELETING), PROC_REF(stop_zooming))
```

File: code/datums/components/scope.dm [hunk 3]
```diff
@@ -126,15 +145,32 @@
 /datum/component/scope/proc/stop_zooming(mob/user)
 	SIGNAL_HANDLER
 
+	if(!zooming)
+		return
+
 	STOP_PROCESSING(SSprojectiles, src)
 	UnregisterSignal(user, COMSIG_MOB_SWAP_HANDS)
+	UnregisterSignal(user, COMSIG_QDELETING)
```
Comment (san7890) [inline code/datums/components/scope.dm:153]:
```suggestion
	UnregisterSignal(user, list(COMSIG_MOB_SWAP_HANDS, COMSIG_QDELETING))
```

yes it is `UnregisterSignal`

File: code/datums/components/scope.dm [hunk 4]
```diff
@@ -101,21 +105,36 @@
...
+		zooming = TRUE
+
+/**
+ * We start zooming by hiding the mouse pointer, adding our tracker overlay and starting our processing.
+ *
+ * Arguments:
+ * * user: The mob we are starting zooming on.
+*/
+/datum/component/scope/proc/zoom(mob/living/user)
 	if(!user.client)
```
Comment (san7890) [inline code/datums/components/scope.dm:124]:
```suggestion
	if(isnull(user.client))
```

may as well

**Review by san7890 — CHANGES_REQUESTED** [review]
small nits, moderate code concerns/considerations

**Pull request merged**